torchserve-vicuna/custom_handler.py

```diff
@@ -0,0 +1,89 @@
+import logging
+from abc import ABC
+
+import packaging.version
+import torch
+import transformers
+from transformers import LlamaTokenizer, LlamaForCausalLM, TextIteratorStreamer
+from threading import Thread
+
+from ts.torch_handler.base_handler import BaseHandler
+from ts.protocol.otf_message_handler import send_intermediate_predict_response
+
+logger = logging.getLogger(__name__)
+logger.info("Transformers version %s", transformers.__version__)
+if packaging.version.parse(torch.__version__) >= packaging.version.parse("2.0.0"):
+    logger.info("PyTorch version is 2.0.0 or greater")
+else:
+    logger.info(
+        "PyTorch version is less than 2.0.0, initializing with meta device needs PyTorch 2.0.0 and greater"
+    )
+
+class ModelHandler(BaseHandler, ABC):
+    def __init__(self):
+        super(ModelHandler, self).__init__()
+        self.initialized = False
+
+    def initialize(self, ctx):
+        self.manifest = ctx.manifest
+        properties = ctx.system_properties
+        model_dir = properties.get("model_dir")
+
+        if torch.cuda.is_available():
+            self.map_location = "cuda"
+            self.device = torch.device(self.map_location)
+        else:
+            self.map_location = "cpu"
+            self.device = torch.device(self.map_location)
+
+        model_path = ctx.model_yaml_config["handler"]["model_path"]
+        seed = ctx.model_yaml_config["handler"]["manual_seed"]
+        dtype_str = ctx.model_yaml_config["handler"]["dtype"]
+        torch.manual_seed(seed)
+
+        dtypes = {"fp32": torch.float32, "fp16": torch.float16, "bf16": torch.bfloat16}
+
+        dtype = dtypes.get(dtype_str, torch.float32)
+        
+        self.model = LlamaForCausalLM.from_pretrained(
+            model_path, 
+            use_cache=False, 
+            max_memory={i: ctx.model_yaml_config["handler"]["max_gpu_memory"] for i in range(ctx.model_yaml_config["handler"]["num_gpus"])},
+            low_cpu_mem_usage=ctx.model_yaml_config["handler"]["low_cpu_mem_usage"],
```

Comment:
Also use load_in_8bit=True for the model and install [bitsandbytes](https://github.com/TimDettmers/bitsandbytes)